File: medium/199-Binary-Tree-Right-Side-View/DFS.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []

        values = []
        stack = [(0, root)]
        while stack:
            depth, node = stack.pop()
            if len(values) == depth:
                values.append(node.val)
            if node.left:
                stack.append((depth + 1, node.left))
            if node.right:
                stack.append((depth + 1, node.right))
        return values

File: medium/199-Binary-Tree-Right-Side-View/DFS.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> values = new ArrayList<>();
        
        if (root == null) {
            return values;
        }
        
        Stack<Pair<Integer, TreeNode>> stack = new Stack<>();
        stack.push(new Pair(0, root));
        
        while (!stack.isEmpty()) {
            Pair<Integer, TreeNode> pair = stack.pop();
            int depth = pair.getKey();
            TreeNode node = pair.getValue();
            
            if (values.size() == depth) {
                values.add(node.val);
            }
            
            if (node.left != null) {
                stack.push(new Pair(depth + 1, node.left));
            }
            
            if (node.right != null) {
                stack.push(new Pair(depth + 1, node.right));
            }
        }
        
        return values;
        
    }
}
